include/sys/dsl_dir.h

```diff
@@ -121,6 +121,11 @@ struct dsl_dir {
 	bplist_t dd_pending_frees;
 	bplist_t dd_pending_allocs;
 
+	kmutex_t dd_activity_lock;
+	kcondvar_t dd_activity_cv;
+	boolean_t dd_activity_cancelled;
+	uint64_t dd_activity_count;
```

Comment:
I think `dd_activity_count` is a count of the number of waiters (which have long holds).  The name might incorrectly imply that this is the count of activities in progress (e.g. is deletion in progress).  How about renaming it to `dd_activity_waiters` or `dd_activity_num_waiters` or `dd_activity_long_holds`?